### PR TITLE
feat(*): add helm lint command 

### PR DESCRIPTION
**Pull request by michelleN**

### CODE REVIEW

File: action/lint.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+//@todo: refactor this method
+func verifyManifests(chartPath string) {
+	manifests, err := manifest.ParseDir(chartPath)
+	if err != nil {
+		log.Die("Error walking manifest files. Err: ", err)
+	}
+
+	for _, m := range manifests {
+		ref, _ := m.VersionedObject.Ref()
```
Comment (technosophos) [inline action/lint.go]: You can now use `Meta()` to get all of the stuff from `Ref()` plus the labels and annotations.

File: action/lint.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		case "Service":
+			s, _ := m.VersionedObject.Service()
+			if s.Name == "" {
+				log.Die("Missing Name in %s", m.Source)
+			}
+
+			val, ok := s.Labels["heritage"]
+			if !ok || (val != "helm") {
+				log.Die("Missing a label: `heritage: helm` in %s", m.Source)
+			}
```
Comment (technosophos) [inline action/lint.go]: Probably if `Kind` is not one of the known kinds we should emit a warning (not an error though). `action.InstallOrder` is probable a cheap and easy way to see if this is a known Kind.

File: action/lint.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+	//assumes chart is in your workspace directory
+	chartPath := filepath.Join(homedir, WorkspaceChartPath, chartName)
+
+	//makes sure all files are in place
+	structure := directoryStructure(chartPath)
+	checkDirectoryStructure(structure, chartPath)
+
+	//checks to see if chart name is unique
+	nameErr := verifyChartNameUnique(chartName)
```
Comment (technosophos) [inline action/lint.go]: Will this get triggered if I update a chart, then lint it?

File: action/lint.go (hunk 1 above)
Comment (michelleN) [inline action/lint.go]: thanks that'll make things much cleaner!

File: action/lint.go (hunk 3 above)
Comment (michelleN) [inline action/lint.go]: Yes.

File: action/lint.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/google/go-github/github"
+	"github.com/helm/helm/chart"
+	"github.com/helm/helm/log"
+	"github.com/helm/helm/manifest"
+	"gopkg.in/yaml.v2"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
```
Comment (sgoings) [inline action/lint.go]: Is there an order for go imports that we should follow? I've seen something like the order of:

- golang builtin libraries
- imported libraries
- internal (to the current project) libraries

Comment (technosophos) [inline action/lint.go]: The recommendation is to split core libraries and "other" libraries.

namespace imports (`. "foo"`) and side effect imports (`_ "foo"`) often are set apart.

I've also seen internal libraries moved to the bottom, but that's not ubiquitous.

Comment (michelleN) [inline action/lint.go]: I've been doing alphabetical order but I would be into what you're suggesting. @technosophos @adamreese any thoughts?

Comment (technosophos) [inline action/lint.go]: Alpha order is also suggested. I usually just split the imports into two groups (core and other), and then let `go fmt` sort them for me.

**Comment by sgoings** [comment]
If I run:

 ~    rm -rf ~/.helm/
 ~    ~/go/src/github.com/helm/helm/bin/helm update
---> Creating /Users/sethgoings/.helm/config.yaml
---> Updating charts
---> Cloning into '/Users/sethgoings/.helm/cache/charts'...
---> Done
 ~    cd ~/.helm/cache/charts/
 ~/.helm/cache/charts  ⑂ master    ~/go/src/github.com/helm/helm/bin/helm lint jenkins
[ERROR] Chart /Users/sethgoings/.helm/workspace/charts/jenkins not found in workspace. Error: stat /Users/sethgoings/.helm/workspace/charts/jenkins: no such file or directory
 ~/.helm/cache/charts  ⑂ master    ~/go/src/github.com/helm/helm/bin/helm fetch jenkins
---> Fetched chart into workspace /Users/sethgoings/.helm/workspace/charts/jenkins
---> Done
 ~/.helm/cache/charts  ⑂ master    ~/go/src/github.com/helm/helm/bin/helm lint jenkins
[WARN] Chart name jenkins already exists in charts repository [github.com/helm/charts].
You might like to change it before continuing
---> Chart [jenkins] has passed all necessary checks

Do we want to be smarter about showing/not showing that WARN message?

**Comment by sgoings** [comment]
Nice!

```
 ~/.helm/workspace/charts/jenkins/manifests    vi /Users/sethgoings/.helm/workspace/charts/jenkins/manifests/jenkins-rc.yaml # removed heritage label
 ~/.helm/workspace/charts/jenkins/manifests    ~/go/src/github.com/helm/helm/bin/helm lint jenkins
[WARN] Chart name jenkins already exists in charts repository [github.com/helm/charts].
You might like to change it before continuing
[ERROR] Missing a label: `heritage: helm` in /Users/sethgoings/.helm/workspace/charts/jenkins/manifests/jenkins-rc.yaml
 ~/.helm/workspace/charts/jenkins/manifests    echo $?
1
```

**Comment by michelleN** [comment]
@sgoings 
Now that I look at it in this context, maybe we should change the message to be something like: "[WARN] Chart name jenkins already exists in charts repository [github.com/helm/charts].
You might want to change it if this is a new chart you are submitting."

Would that make things more clear or were you thinking we should do something else?

**Comment by technosophos** [comment]
@sgoings we should probably add a global -q flag that will suppress INFO and WARN messages for all commands. Thoughts?

**Comment by technosophos** [comment]
@sgoings @michelleN Would it be a better user experience to have `helm lint` take a path instead of a chart name? Then we could lint without fetching? e.g.:

```
$ helm lint $HELM_HOME/cache/foo/bar
$ helm lint ./bar
$ helm lint /some/other/path
```

**Comment by sgoings** [comment]
@michelleN my thought was that if I'm editing the jenkins chart (out of helm/charts)... shouldn't it be ok that the jenkins chart already exists in helm/charts?

**Comment by michelleN** [comment]
@technosophos yeah I was playing with the idea of taking in a path instead. I think it'll be less confusing for sure. @sgoings what do you think?

**Comment by sgoings** [comment]
@technosophos I agree - but this is @michelleN's first pass... the current contributor workflow is all workspace centric:

```
Fork charts repository
helm create foo && helm edit foo
cd ~/.helm/cache && git checkout -b foo
git add foo && git commit && git push
Submit a Pull Request
```

all we're doing is adding a:

```
helm lint foo
```

in the middle of that process.

Why not both? (in a future effort + issue)?

**Comment by michelleN** [comment]
@sgoings I like the idea of opening up an issue and making it a near future effort. +1

**Comment by michelleN** [comment]
@sgoings Yes, it should be okay to edit the jenkins chart (for example). That's why I made that a warning instead of an error. I do still want the user to be cautious if they're planning on making a new or variant chart out of an existing one. I guess I don't know exactly what the user workflow would like for lint and there are so many various use cases for lint (like in the context of CI/automatically merging a chart into the charts repo, etc.)

**Comment by sgoings** [comment]
@michelleN - just an item for discussion. In the ideal case, helm lint should probably be smart enough to catch the case where you're editing a pre-existing chart. Not a blocker for merging this work though.

**Comment by michelleN** [comment]
@sgoings I agree. I'll put it as an acceptance criteria for round 2 if that's cool with everyone. @technosophos

**Comment by technosophos** [comment]
I'm good with delaying both of those until after 0.2

**Comment by michelleN** [comment]
@sgoings I went ahead and modified the warning message to make it less hostile for now

File: action/lint.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	"github.com/helm/helm/log"
+	"github.com/helm/helm/manifest"
+	"gopkg.in/yaml.v2"
+)
+
+// Owner is default Helm repository owner or organization.
+var Owner = "helm"
+
+// Project is the default Charts repository name.
+var Project = "charts"
```
Comment (arschles) [inline action/lint.go]: can this and the above be constants?

Comment (michelleN) [inline action/lint.go]: yeah that's probably a good idea.

**Pull request merged**